src/control/server/storage_scm_prep.go

```diff
@@ -24,335 +24,16 @@
-	}
-
-	return true, nil // memory allocation reset requires a reboot
-}
-
-func (ps *prepScm) removeNamespace(devName string) (err error) {
-	ps.log.Infof("removing SCM namespace, may take a few minutes...\n")
-
-	_, err = ps.runCmd(fmt.Sprintf(cmdScmDisableNamespace, devName))
-	if err != nil {
-		return
-	}
-
-	_, err = ps.runCmd(fmt.Sprintf(cmdScmDestroyNamespace, devName))
-	if err != nil {
-		return
-	}
-
-	return
-}
-
-// hasFreeCapacity takes output from ipmctl and checks for free capacity.
-//
-// external tool commands return:
-// $ ipmctl show -d PersistentMemoryType,FreeCapacity -region
-//
-// ---ISetID=0x2aba7f4828ef2ccc---
-//    PersistentMemoryType=AppDirect
-//    FreeCapacity=3012.0 GiB
-// ---ISetID=0x81187f4881f02ccc---
-//    PersistentMemoryType=AppDirect
-//    FreeCapacity=3012.0 GiB
-//
-// FIXME: implementation to be replaced by using libipmctl directly through bindings
-func hasFreeCapacity(text string) (hasCapacity bool, err error) {
-	lines := strings.Split(text, "\n")
-	if len(lines) < 4 {
-		return false, errors.Errorf("expecting at least 4 lines, got %d",
-			len(lines))
-	}
-
-	for _, line := range lines {
-		entry := strings.TrimSpace(line)
-
-		kv := strings.Split(entry, "=")
-		if len(kv) != 2 {
-			continue
-		}
-
-		if kv[0] == "PersistentMemoryType" && kv[1] == "AppDirect" {
-			hasCapacity = true
-			continue
-		}
-
-		if kv[0] != "FreeCapacity" {
-			continue
-		}
-
-		if hasCapacity && kv[1] != "0.0 GiB" {
-			return
-		}
-
-		hasCapacity = false
-	}
-
-	return
-}
-
-func parsePmemDevs(jsonData string) (devs []pmemDev, err error) {
-	// turn single entries into arrays
-	if !strings.HasPrefix(jsonData, "[") {
-		jsonData = "[" + jsonData + "]"
-	}
-
-	err = json.Unmarshal([]byte(jsonData), &devs)
-
-	return
-}
-
-// createNamespaces runs create until no free capacity.
-func (ps *prepScm) createNamespaces() ([]pmemDev, error) {
-	devs := make([]pmemDev, 0)
-
-	for {
-		ps.log.Infof("creating SCM namespace, may take a few minutes...\n")
-
-		out, err := ps.runCmd(cmdScmCreateNamespace)
-		if err != nil {
-			return nil, errors.WithMessage(err, "create namespace cmd")
-		}
-
-		newDevs, err := parsePmemDevs(out)
-		if err != nil {
-			return nil, errors.WithMessage(err, "parsing pmem devs")
-		}
-		devs = append(devs, newDevs...)
-
-		state, err := ps.GetState()
-		if err != nil {
-			return nil, errors.WithMessage(err, "getting state")
-		}
-
-		switch state {
-		case types.ScmStateNoCapacity:
-			return devs, nil
-		case types.ScmStateFreeCapacity:
-		default:
-			return nil, errors.Errorf("unexpected state: want %s, got %s",
-				types.ScmStateFreeCapacity.String(), state.String())
-		}
-	}
-}
-
-func (ps *prepScm) GetNamespaces() ([]pmemDev, error) {
-	if err := ps.checkNdctl(); err != nil {
-		return nil, err
-	}
-
-	out, err := ps.runCmd(cmdScmListNamespaces)
-	if err != nil {
-		return nil, err
-	}
-
-	return parsePmemDevs(out)
-}
+// core scm prep code moved to storage/scm/ipmctl.go
```

Comment:
Any reason why the interface shouldn't go with it?

Reply:
I tried to rein myself in from changing too much at once in one PR. The primary goal of this work was to get the core SCM code isolated in its own package and decoupled from the RPC/server stuff. Leaving the interface here for now reduces the review burden while the rest of the work shapes up.

Reply:
That's fair. Might be a good idea to add a TODO if it's going somewhere else eventually.

Reply:
Done.